texastribune/salesforce-stripe
# Charge new donations immediately

#### What's this PR do?

1. Charges new donations immediately instead of waiting for the batch job to run.
1. Makes some minor PEP8 changes.
1. Checks for duplicate transactions.

#### Why are we doing this? How does it help us? 

The batch job was designed to run once or a few times daily. But at certain times (like membership drives) they need to have more up-to-date numbers. So we run the batch job more frequently. But we can only configure it to run at most once an hour. Moving charging for single donations and the initial charge for new recurring donations to the time of submission means those numbers are always up-to-date without having to run the batch job frequently. 

#### How should this be manually tested?

1. Run the project as normal. 
1. Make some test transactions. Try regular, recurring, Circle and Blast. 
1. Make sure you see a message about charging the card and a successful call to the Stripe API for it.
1. Try new donations with and without `CHECK_FOR_DUPLICTES` set and check that it behaves as expected.

#### Have automated tests been added?

No, but existing ones have been modified.

#### Has this been tested on mobile?

N/A

#### Are there performance implications?

This will make it take longer to process a transaction. But since it's done asynchronously in the background it won't mean any additional time for the user.

#### What are the relevant tickets?

Off-sprint.

#### How should this change be communicated to end users?

Let membership know that numbers will be up-to-date all the time.

#### Next steps?

Add the new `CHECK_FOR_DUPLICATES` to our test and production configuration.

#### Smells?

We don't really have a way to know if the user intended to donate twice in the same day. So it's possible that they will intentionally do so. But we'll get an alert and can contact the user and process it manually in those cases. And this will allow charges for different types in the same day: a Blast and a membership will go through. But for two recurring donations in the same day, the second will be stopped. But this behavior can be disabled with a new variable: `CHECK_FOR_DUPLICATES`.

#### Has the relevant documentation/wiki been updated?

Not yet.
